### PR TITLE
docs: Expand loader layer readme

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add more details clarifying the Loader class vs the loader layer.

Since the first think the linked document talks about includes "access to
Fluid storage as well as consensus over a quorum of clients" it was pretty unclear how this relates to loading.

### CODE REVIEW

File: packages/loader/README.md [hunk 1]
```diff
@@ -1,3 +1,12 @@
 # Fluid Loader
 
-For an overview of the Fluid Loader, please refer to the [@FluidFramework/container-loader package README](container-loader/README.md).
+This is about the "loader layer" of the FLuid Framework, not to be confused with the `Loader` class (which is part of this layer).
```
Comment (ghoshkaj) [inline packages/loader/README.md:3]:
```suggestion
This is about the "loader layer" of the Fluid Framework, not to be confused with the `Loader` class (which is part of this layer).
```

File: packages/loader/README.md [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
+
+The `loader` layer contains code which gets directly included into fluid applications.
+This this layer is kept minimal to reduce the the need to update the application when the Fluid Framework changes,
```
Comment (ghoshkaj) [inline packages/loader/README.md:6]:
```suggestion
This layer is kept minimal to reduce the need to update the application when the Fluid Framework changes,
```

Comment (justus-camp) [inline packages/loader/README.md:6]:
```suggestion
This layer is kept minimal to reduce the the need to update the application when the Fluid Framework changes,
```

Comment (justus-camp) [reply packages/loader/README.md:6]: Looks like @ghoshkaj beat me to it 😄

File: packages/loader/README.md [hunk 3]
```diff
@@ -1,3 +1,12 @@
...
+as well as keep the size application small.
+Instead most of the code is placed in either the `runtime` or `driver` layers and dynamically loaded by the `loader`.
+
+The loading is initiated by loading a container, which involves the appropriate drivers and runtime for the container.
```
Comment (ghoshkaj) [inline packages/loader/README.md:10]: Something about "The loading is initiated by loading a container" strikes me as odd but I don't know of a better way to say it. 🤔

**Comment by ChumpChief** [comment]
Thanks for taking a look at this.  Part of the challenge is that it's a weirdly-defined layer.  Might be interesting to clarify how much you want to talk about "what the layer is" vs. "what the layer should be".  Disclaimer: I don't really think this layer should exist.

IMO the container-loader package might be the only one that makes sense in this layer.  The other packages seem mostly like things that can be passed into the loader layer and/or grab-bag, and probably don't belong there.  I suspect this is part of why the current README just says "just go look at container-loader"

Within the container-loader package, only a few things make sense in the layer as well.  We've talked for a while about wanting to move the DeltaManager to the runtime, for example.  I could see quorum and audience moving there too someday.  Basically trying to get the loader layer out of the op processing flow entirely.

With this in mind, maybe consider a core message of "The loader layer is directly consumed by hosts, and its main purpose is to create and load containers which in turn hold the runtime layer with the real data.  It also currently contains a portion of op processing logic that sits between between the driver and runtime."

**Comment by CraigMacomber** [comment]
> The other packages seem mostly like things that can be passed into the loader layer and/or grab-bag, and probably don't belong there. I suspect this is part of why the current README just says "just go look at container-loader"

If something is passed into the loader layer, then its types have to be defined in a place the loader layer can depend on. In my understanding of our current layering that means they have to be in the loader layer.

Reading build-tools/packages/build-tools/data/layerInfo.json, I find 4 Definition layers, 8 Util layers and 15 Client layers and a couple others. Even based on that, I don't see a place in the codebase to put client only types that could be passed into the loader  (from either the application, the driver or the runtime) other than in the loader layer. Perhaps they could have their definitions and implementations separated, and the implementations (ex the DeltaManager Class) moved into the runtime or drivers?

Even if that separation was done, those abstractions would still be defined and documented in this layer.

That said, unless someone has some specific text about the future of this layer they want added to this PR before this is merged, I'd like to scope refactoring out of this PR. I'd also like someone to fix the container-loader" readme since its confusing, but thats also a separate issue.

**Comment by CraigMacomber** [comment]
> The loader layer is directly consumed by hosts, and its main purpose is to create and load containers which in turn hold the runtime layer with the real data. It also currently contains a portion of op processing logic that sits between between the driver and runtime.

I find the term "read data" insufficiently descriptive (the loader is real data, as are the driver and runtime packages it loads, as is the source code, application data etc). Also the runtime layer is a code organization concept so I don't get what it means for the container to hold it. Also "it" in the second sentences isn't clear what is is referring to. 

Also, readers mind read this doc with little context about fluid specific stuff, so I don't think "hosts" is super clear: there are many hosts on the web.

Anyway, that description seems focused on container-loader and not the loader layer. Feel free to go update container-loader'  readme if you think you can make it more clear (I find it very confusing right now), which will remove a lot of the requirements of this readme which is really just about what code is in the folder.

**Comment by ChumpChief** [comment]
> If something is passed into the loader layer, then its types have to be defined in a place the loader layer can depend on. In my understanding of our current layering that means they have to be in the loader layer.

Not really - for example, `IDocumentServiceFactory` is defined in driver-definitions and the implementations are from the driver layer, and these are passed into the loader.  In any case, my point is really just that looking at the other stuff besides container-loader is sort of a distraction from the meat of the loader layer.  I'm not suggesting you refactor anything in this PR - just saying that if we had appropriate layering, then the term "loader layer" would be equal to "container-loader" IMO.

Re: "real data" - I'm referring to a customer's perspective.  Their "real data" is the e.g. the map they created and the key/values they stored in it, as opposed to all the "loader", "container", "context" things we have hanging around.  This "real data" only becomes a concept down in the runtime layer.

Re: "hosts" - "Hosts" (a.k.a. "apps") are one of our two customers (the other being "container authors").  We may not have this defined well enough elsewhere, but you can't really talk about the loader layer without knowing what it means - this layer was built specifically for them.

**Pull request closed without merging**